### ISSUE TITLE
Allow key-spacing to align separated groups as well

### ISSUE THREAD

**Issue by apierzch** [eslint/eslint#2668]
Currently example in documentation says that separated groups will not align:
```
var obj = {
    a:    value,
    bcde: 42,

    fg: foo(),
    h:  function() {
        return this.a;
    },
    ijkl: 'Non-consecutive lines form a new group'
};
```

It would be useful to have option that would enforce or allow such alignment, so that I can have: 
```
var obj = {
    a:    value,
    bcde: 42,

    fg:   foo(),
    h:    function() {
        return this.a;
    },
    ijkl: 'Non-consecutive lines form a new group'
};
```

Same for aligning on colon.

I'll try to provide a PR if you agree with me.

**Comment by ilyavolodin** [comment]
If you want all of those properties to be aligned, why do you separate them out with a line break? Wouldn't it be easier to just remove line break?

**Comment by nzakas** [comment]
I'm sorry, I don't understand what you're asking, can you be more explicit?

**Comment by apierzch** [comment]
@ilyavolodin the example is oversimplified. The same problem appears when one property expands to multiline, as in this example:
```
var obj = {
  a :      value,
  longer : value,
  multi :  {
    sth : value
  },
  x :      value
};
```
In this example value of `x` is aligned to value of `longer`, but eslint treats it as a new group and reports error. For me this is undesired.

**Comment by btmills** [comment]
@heldinz encountered this in #2674, so I'll merge that discussion into this one. The example given there was:

```js
{
    title:       'My title',
    description: 'My description',
    options:     {
        optionOne:   'First option',
        optionTwo:   'Second option',
        optionThree: 'Third option'
    },
    isDefault:   true
}
```

Where currently `isDefault` currently errors with too much space because it's considered a separate group.

For me, the compelling argument in #2674 is that this is the style IntelliJ provides, and we don't support that.

To quote my own suggestion:

> Should there be an option that would match IntelliJ's behavior? "one-group": true perhaps?

And the response from @heldinz:

> That would certainly work for me. Or perhaps some sort of "strictness" option regarding grouping? Something like "allow-newlines-in-group": true ?

Thoughts?

**Comment by apierzch** [comment]
Completely agree with your suggestion @btmills 

My reasoning is actually the same - I use IntelliJ as well.

**Comment by nzakas** [comment]
@btmills this is your baby, I trust your instincts

**Comment by btmills** [comment]
We should support this use case, but I'm not sure how. Should this be the default behavior, with groups being an option, or are groups the default, with this being optional? Either way, what does the option look like?

**Comment by heldinz** [comment]
I think groups in general are a good idea and should remain the default behaviour. I can't imagine visually separating groups with blank lines and still expecting the values to line up. So how about the current behaviour (a new group is indicated by a blank line or a multi-line property) as default, since that seems to be fine for most people, with the option being to change what indicates a new group (i.e. still blank lines but not multi-line properties), e.g. my `allow-newlines-in-group suggestion` from earlier, or perhaps `allow-mutliline-props-in-group`, or even something like `intellij-style-grouping`?

It's worth noting that what IntelliJ does is almost like nesting of groups: the values within the "options" object in the example above are aligned with each other, while the value of the options key itself (the opening curly brace of the object) is aligned with the values above and below it. I'm not sure of the best way to communicate all of this expected behaviour in one option name (let alone how difficult it is to implement)... Maybe an option like `intellij-style-grouping` would make the most sense, if it's well-documented what it actually does.

**Comment by IanVS** [comment]
For determining the default behavior, I think it's necessary to follow the principal of least surprise. Is it more surprising for an empty line or multi line to create a new group (and thereby a new level of alignment), or for all values within an object to follow the "align" setting specified as the option? I like the ability to beak up into groups with their own alignment, but I have to think that behavior is surprising if you didn't know that is what is happening when you add an empty line.

**Comment by nzakas** [comment]
The principle of least surprise applies only to things that are new. For existing things, the burden of proof shifts to what will cause the least pain to existing users.

**Comment by zyklus** [comment]
My only concern is that we don't _disallow_ mis-aligned groups with any option.  i.e. I'm good with an option that _allows_ different groups to be aligned, but you should be able to mix styles, e.g. I want this code to be valid:

```js
a   : 10
b   : 10

foo : 10
bar : 10

c : 10
d : 10

// this group would probably be invalid since it doesn't align
// to anything and has too much whitespace
e  : 10
f  : 10
```

**Comment by gyandeeps** [comment]
ping @eslint/eslint-team

**Comment by btmills** [comment]
@zyklus mixing styles sounds like a separate request. Can you file a new issue if you still want that?

To summarize my understanding: The goal here is to support IntelliJ's indent style by aligning all properties of an object to the same level, regardless of their vertical separation. In terms of the implementation, this will consist of a new option that maintains the current default behavior when unset and treats all properties of an object as part of the same group when set. That should be trivial, so all that's left is a decision on the option. One idea is `"groups": false` (defaults to `true`), but I don't think the name is intuitive here; it's more a reflection of the internal implementation.

**Comment by zyklus** [comment]
@btmills - I think it would be easier to implement it at the same time though.  My basic suggestion is that if `groups: false` is set that the logic becomes:

```js
if (alignedToPrevGroup || alignedToNextGroup) {
```

Possibly more like: `groups = [all, any, neighbors]` (again, naming up for discussion but hopefully the meaning is clear enough?)

**Comment by btmills** [comment]
For the original issue, how about `"groups": ["all", "adjacent"]`? @apierzch @heldinz :+1: / :-1: ?

@zyklus I'm not sold yet, but I could be convinced. What you're asking for is that there are three possible ways to align a property:

1. Whatever the rule calculates given `align`, `beforeColon`, and `afterColon`
2. Same as the previous property
3. Same as the next property

Would you require that at least one property at each alignment level meets the requirements for (1), or could there be arbitrary alignment levels provided there's more than one property at that level?

Some examples of what you want would help.

**Comment by apierzch** [comment]
Actually `all` vs `adjanced` doesn't sound intuitive for me. In my opinion `"groups" : true|false` seems to reflect expected behavior (and I'm not influenced by implementation, since I don't know it).

**Comment by zyklus** [comment]
@btmills -- To me the issue is newlines/comments and when/where they're added.

Let's say I have:

```js
{
  foo    : 123,
  bar    : 456,
  foobar : 789,
  foobaz : 012
}
```

All nice and aligned.  Now I want to add a comment:

```js
{
  foo    : 123,
  bar    : 456,
  
  // combo options
  foobar : 789,
  foobaz : 012
}
```

I don't want to be forced to re-align the first two keys.  At the same time, I'd copied these objects from one object to another, e.g. it was originally:

```js
{
  foo : 123,
  bar : 456
}
  
// combo options
{
  foobar : 789,
  foobaz : 012
}
```

I don't want to be forced to align keys.  What it comes down to is that I'm perfectly happy with different methods of alignment here, but am not happy with chaos:

```js
{
  foo : 123,
  foobar : 789,
  bar : 456
  foobaz : 012
}
````

When it comes down to it, I don't really care about the "neighbor" option -- What I care about is that it doesn't complain about having _too much_ whitespace before a colon when the colons are aligned.

**Comment by btmills** [comment]
@zyklus:

> What I care about is that it doesn't complain about having *too much* whitespace before a colon

Ah! Take a look at #3549. It adds a `mode` option, which defaults to `"strict"`, the current behavior, but can be set to `"minimum"`, which will allow one or more spaces when `beforeColon` or `afterColon` are set to `true`.

**Comment by heldinz** [comment]
@btmills 

> To summarize my understanding: The goal here is to support IntelliJ's indent style by aligning all properties of an object to the same level, regardless of their vertical separation.

That's not *100%* correct. IntelliJ does recognise different alignment groups within an object's properties if those properties are separated by a blank line, e.g. here the required properties are aligned as one group while the optional properties are aligned in another:

    propTypes: {
        requiredString: ReactPropTypes.string.isRequired,
        requiredNumber: ReactPropTypes.number.isRequired,

        optionalBool: ReactPropTypes.bool
    }

The difference between IntelliJ's style and the current eslint implementation is that in eslint, line breaks are enough to trigger a new alignment group, while IntelliJ requires actual blank space between the groups.

> For the original issue, how about `"groups": ["all", "adjacent"]?` @apierzch @heldinz :+1: / :-1: ?

I don't find the suggestion too intuitive either. Given the difference in style I described here, how about something more like `"grouping": ["line-break", "vertical-space"]`?

**Comment by zyklus** [comment]
@heldinz -- I think that those options make even less sense, but that's just my opinion.

Perhaps just a boolean?  `multipleGroups: true|false`?

**Comment by heldinz** [comment]
@zyklus I'm not thrilled with the names either, and I'd also be happier with a boolean option, if we can find one that's sensible. But since the purpose of this enhancement is to distinguish *how* groups are defined, not *whether* or not we have groups at all, boolean options like `multipleGroups: true|false` don't seem to me to be descriptive enough.

>> while IntelliJ requires actual blank space between the groups.

> Is this a reference to comments?

Not in particular, it's just stating IntelliJ's criteria for defining alignment groups. (For me personally, multiline values are a much more common issue than comments.) IntelliJ will only consider adjacent properties to be separate alignment groups if they are separated by actual blank lines/vertical space. Anything else coming between them, including comments and [multiline values][1] (described earlier in this issue), is not enough for IntelliJ to consider what follows a separate alignment group.

For example, IntelliJ considers this to be one alignment group. eslint currently does not, because of the multiline value:

    {
        title:       'My title',
        description: 'My description',
        getTitle:    function () {
            return this.title;
        },
        isDefault:   true // IntelliJ aligns with the above properties, eslint does not
    }

But put a blank line between the properties, and IntelliJ will consider them separate alignment groups, just as eslint does:

    {
        title:       'My title',
        description: 'My description',
        getTitle:    function () {
            return this.title;
        },

        isDefault: true // IntelliJ does not align with the above properties, neither does eslint
    }

To repeat my understanding of the problem: @btmills would like to keep the current eslint behaviour as the default, but support the IntelliJ style too. So the question is, how do we define the option that says we want to use IntelliJ's criteria for defining alignment groups instead of eslint's?

[1]:   https://github.com/eslint/eslint/issues/2668#issuecomment-108496087

**Comment by zyklus** [comment]
>  But since the purpose of this enhancement is to distinguish how groups are defined, not whether or not we have groups at all, boolean options like multipleGroups: true|false don't seem to me to be descriptive enough.

I would argue that that's a matter of perception.  I view this as "everything is the same" or "separated into groups", so `group: true` makes sense to me.  I'm not sure what the general consensus on this would be though.

**Comment by heldinz** [comment]
Ok, maybe @btmills can clarify?

**Comment by btmills** [comment]
For now I'll make sure I understand correctly, then once I'm clear on that we can move on to option format.

I've seen discussed three possible modes:

1. Groups of properties separated by new lines, so a multi-line function expression value would be the end of a group. This is the current default behavior, and it needs to remain so since we don't want this to be a breaking change.

    ```js
    ({
        method: function () {
            return 42;
        },
        hi: 'name',

        foo: 'hello',
        b:   'world'
    });
    ```

1. Groups of properties separated by empty lines (no code, no comments, just whitespace). Unlike the current behavior, multi-line values continue groups instead of ending them. This is the grouping behavior exhibited by IntelliJ.

    ```js
    ({
        method: function () {
            return 42;
        },
        hi:     'name',

        foo: 'hello',
        b:   'world'
    });
    ```

1. No notion of groups at all / one giant group. All properties in an object are aligned to the same level.

    ```js
    ({
        method: function () {
            return 42;
        },
        hi:     'name',

        foo:    'hello',
        b:      'world'
    });
    ```

First, **are all three descriptions correct?** If so, then second, **do we need to add support for 2, 3, or both?**

**Comment by zyklus** [comment]
@btmills - I have a 4th, and I'm sure someone not on this thread has a 5th and 6th.

My suggestion:  Implement option 2, along with #3549.  This handles cases 3 & 4, along with most other weird cases by putting the linter into "loose" mode

**Comment by heldinz** [comment]
@btmills - That fits my understanding for 1 and 2. I can imagine there being a use case for 3 as well, but it's not a priority for me personally.

**Comment by apierzch** [comment]
@btmills Yes, all descriptions are correct. I am mostly interested in option *3*, since this is how IntelliJ Idea aligning works.

**Comment by zyklus** [comment]
@apierzch -- Out of curiosity, would option 2 along with "loose" spacing rules before/after colons work for you?  Or is not having errors on mis-alignment a deal-breaker?

i.e. both of these would pass and it would be up to you to enforce intra-group alignment:

```js
{
  foo    : 1,
  bar    : 2,

  fooBar : 3,
  fooBaz : 3
}
```

```js
{
  foo      : 1,
  bar      : 2,

  fooBar : 3,
  fooBaz : 3
}
```

But mis-aligned keys within a group would error

**Comment by heldinz** [comment]
@apierzch 

> I am mostly interested in option 3, since this is how IntelliJ Idea aligning works.

Really? I am interested in option 2 because that is the behaviour I experience with IntelliJ Idea.

**Comment by apierzch** [comment]
@heldinz I have just checked - if you add an empty line, it creates a new group in IntelliJ Idea (I thought same as you before, so it was a surprise for me):
```
var myLink = {
        img   : "btn.gif",
        text  : "Button",
        width : 128,

        longValue : 1
    }
```

@zyklus I think I cannot achieve such alignment with IntelliJ Idea, so I'd like to have an error if something like it happens, so that I do not commit just by reformatting

**Comment by ilyavolodin** [comment]
@btmills Should we mark this as accepted or do you want to flash out options first?

**Comment by btmills** [comment]
What I'm seeing from that discussion is that we don't need to be able to turn off grouping entirely (option 3), we just need to be able to change how groups are determined (options 1 and 2). With that in mind, how about a new option `"grouping"` which has two possible values:

- `"multi-line"` (default): Values spanning multiple lines end their group
- `"empty-line"`: Blank lines (no comments, no code, just whitespace) create new groups

@ilyavolodin let's hold off - it's going to be accepted, but I don't want someone to sink time implementing it, only to have the options change later.

**Comment by heldinz** [comment]
@btmills That sounds great to me! :+1:

**Comment by btmills** [comment]
It's been a week with no objections, so let's go with this.

**Comment by mourner** [comment]
@btmills just to clarify, would your proposal cover the following use case?

```js
// both aligned and non-aligned groups are accepted:
var foo = {
    // not aligned
    first: 1,
    second: 2,

    // aligned
    third:  3,
    fourth: 4
}

// but you can't mix styles in a block like this:
var foo = {
	first: 1,
	second: 2,
	third:  3,
	fourth: 4
};
```

**Comment by btmills** [comment]
@mourner no, please file a separate issue if you'd like to discuss that.

**Comment by mourner** [comment]
@btmills thanks, created #3938.

**Comment by theage** [comment]
+1 for #2674 - alignment should continue after multi-line properties without empty / whitespace-only lines to separate groups.

```javascript
{ // wanted
    alpha: 1,
    beta : {
    },
    c    : 3
};

{ // who would want this?!
    alpha: 1,
    beta : {
    },
    c: 3
};
```

**Comment by nzakas** [comment]
Unfortunately, it looks like no one was willing to work on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that "help wanted" issues that aren't addressed within 90 days tend to never get addressed, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to and not something that members of community were willing to implement.
